File: Legacy/Array/560.subarray_sum_equals_k.py
from typing import List

class Solution:
    # O(n^2) solution
    # def subarraySum(self, nums: List[int], k: int) -> int:
    #     n = len(nums)
    #     if n == 0:
    #         return 0
    #     running = [0]
    #     for i in range(n):
    #         running.append(running[i] + nums[i])
    #     ans = 0
    #     for i in range(n):
    #         for j in range(i, n):
    #             if running[j+1] - running[i] == k:
    #                 ans += 1
    #     return ans
                
        def subarraySum(self, nums: List[int], k: int) -> int:
            ans = 0
            n = len(nums)
            if n == 0:
                return ans
            table = {0: 1}
            running = 0
            for i in range(n):
                running += nums[i]
                residual = running - k
                if residual in table:
                    ans += table[residual]
                if running in table:
                    table[running] += 1
                else:
                    table[running] = 1
            return ans

File: C++/Hash Table/560. Subarray Sum Equals K.cpp
// 类似于2sum的思路，用哈希表保存前缀和出现的次数
// 计算到新的前缀和x的时候，去哈希表寻找之前等于x-k的前缀和出现次数

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> sumCount;
        sumCount[0] = 1;
        int prefixSum = 0;
        int ans = 0;
        for (int n : nums) {
            prefixSum += n;
            int needed = prefixSum - k;
            if (sumCount.count(needed)) {
                ans += sumCount[needed]; 
            }
            sumCount[prefixSum]++;
        }
        return ans;
    }
};